src/extensions/agent_modules/cloud_communication/eg_mqtt_broker_client/src/adu_enrollment_management.c

```diff
@@ -0,0 +1,446 @@
+#include "aduc/logging.h"
+#include "aduc/string_c_utils.h" // IsNullOrEmpty
+#include "aducpal/time.h" // time_t
+
+#include "parson.h" // JSON_Value, JSON_Object
+
+#include <mosquitto.h> // mosquitto related functions
+#include <mqtt_protocol.h>
+#include <string.h>
+
+// Default timeout for enrollment request message - 1 hour.
+// This can be overridden by the specifying value in du-config.json
+// e.g.,
+//   "agent.enrollmentSettings.enrollmentRequestTimeoutSeconds": 3600
+#define DEFAULT_ENR_REQ_TIMEOUT_SECONDS (60 * 60)
+
+// Default minimum retry delay for enrollment request message - 5 minutes.
+#define DEFAULT_ENR_REQ_MIN_RETRY_DELAY_SECONDS (60 * 5)
+
+// Enrollment request message content, which is always an empty JSON object.
+static const char* s_enr_req_message_content = "{}";
+static const size_t s_enr_req_message_content_len = 2;
+
+#define CONST_TIME_5_SECONDS 5
+#define CONST_TIME_30_SECONDS 30
+
+#if _ADU_DEBUG
+time_t s_lastHeartBeatLogTime = 0;
+#endif
+
+/**
+ * @struct ADU_ENROLLMENT_DATA_TAG
+ * @brief Struct to represent enrollment status management data for an Azure Device Update (ADU) enrollment.
+ */
+typedef struct ADU_ENROLLMENT_DATA_TAG
+{
+    /// Current enrollment state.
+    ADU_ENROLLMENT_STATE enrollmentState;
+
+    /// Time of the last attempt to enroll.
+    time_t enr_req_lastAttemptTime;
+
+    /// Time of the last enrollment error.
+    time_t enr_req_lastErrorTime;
+
+    /// Time of the last successful enrollment.
+    time_t enr_req_lastSuccessTime;
+
+    /// Time of the next enrollment attempt.
+    time_t enr_req_nextAttemptTime;
+
+    /// Timeout duration in seconds for enrollment attempts.
+    unsigned int enr_req_timeOutSeconds;
+
+    /// Minimum delay in seconds before retrying enrollment.
+    unsigned int enr_req_minRetryDelaySeconds;
+
+    /// Correlation ID for the enrollment request.
+    char* enr_req_correlationId;
+
+    /// Time of the last received enrollment response.
+    time_t enr_resp_lastReceivedTime;
+
+    /// Device update instance associated with the enrollment response.
+    char* enr_resp_duInstance;
+
+    /// Result code indicating the enrollment response status.
+    int enr_resp_resultCode;
+
+    /// Extended result code providing additional enrollment response details.
+    int enr_resp_extendedResultCode;
+
+    /// Description of the enrollment response result.
+    char* enr_resp_resultDescription;
+} ADU_ENROLLMENT_DATA;
+
+static ADU_ENROLLMENT_DATA s_enrMgrState = { 0 }; //!< Enrollment state
+
+/**
+ * @brief A helper function use for setting s_enrMgrState.enr_req_correlationId.
+ */
+static void SetCorrelationId(const char* correlationId)
+{
+    if (s_enrMgrState.enr_req_correlationId != NULL)
+    {
+        free(s_enrMgrState.enr_req_correlationId);
+    }
+
+    if (correlationId == NULL)
+    {
+        s_enrMgrState.enr_req_correlationId = NULL;
+    }
+    else
+    {
+        s_enrMgrState.enr_req_correlationId = strdup(correlationId);
```

Comment:
only free like lin 99 if this succeeds in a tmp var then commit. fn should probably return bool failure if oom